File: backend/src/utils/content_parser.py
import re
from typing import List, Dict, Any


def parse_content_with_markers(content: str) -> Dict[str, Any]:
    """
    Parse content and extract information about personalization markers
    """
    result = {
        'content_without_markers': content,
        'markers_found': [],
        'marker_count': 0
    }
    
    # Regular expression to match personalization markers
    # Pattern: {{personalize:marker_type}}content{{personalize:end}}
    marker_pattern = r'\{\{personalize:([^}]+)\}\}([\s\S]*?)\{\{personalize:end\}\}'
    
    matches = re.findall(marker_pattern, content)
    result['marker_count'] = len(matches)
    
    # Store information about each marker
    for marker_type, marker_content in matches:
        result['markers_found'].append({
            'type': marker_type,
            'content': marker_content.strip()
        })
    
    return result


def replace_markers_in_content(content: str, user_profile: Dict[str, Any]) -> str:
    """
    Replace personalization markers in content based on user profile
    """
    # Regular expression to match personalization markers
    marker_pattern = r'\{\{personalize:([^}]+)\}\}([\s\S]*?)\{\{personalize:end\}\}'
    
    def replace_marker(match):
        marker_type = match.group(1)
        marker_content = match.group(2)
        
        # Check if this marker should be shown based on user profile
        if should_show_marker(marker_type, user_profile):
            return marker_content
        else:
            return ''  # Remove the content if marker shouldn't be shown
    
    # Replace all markers in the content
    result_content = re.sub(marker_pattern, replace_marker, content)
    
    return result_content


def should_show_marker(marker_type: str, user_profile: Dict[str, Any]) -> bool:
    """
    Determine if a marker should be shown based on user profile
    """
    # Check experience level
    if marker_type == user_profile.get('experience_level', ''):
        return True

    # Handle experience level ranges (e.g., if user is 'intermediate', show 'beginner' content too)
    experience_level = user_profile.get('experience_level', '').lower()
    if experience_level == 'intermediate' and marker_type == 'beginner':
        return True
    if experience_level == 'advanced' and (marker_type == 'beginner' or marker_type == 'intermediate'):
        return True

    # Check software background
    software_background = user_profile.get('software_background', [])
    for sb in software_background:
        # Normalize the software name to match marker types (e.g., 'Python' -> 'python', 'python-focused')
        normalized_sb = sb.lower().replace(' ', '-').replace('_', '-')
        if marker_type in normalized_sb or normalized_sb in marker_type:
            return True
        # Check for partial matches
        if marker_type.replace('-', '').replace('_', '') in normalized_sb.replace('-', '').replace('_', ''):
            return True

    # Check hardware background
    hardware_background = user_profile.get('hardware_background', [])
    for hb in hardware_background:
        # Normalize the hardware name to match marker types
        normalized_hb = hb.lower().replace(' ', '-').replace('_', '-')
        if marker_type in normalized_hb or normalized_hb in marker_type:
            return True
        # Check for partial matches
        if marker_type.replace('-', '').replace('_', '') in normalized_hb.replace('-', '').replace('_', ''):
            return True

    # Special case: marker type might be a combination like "python-raspberry-pi"
    if '-' in marker_type:
        parts = marker_type.split('-')
        user_backgrounds = [
            sb.lower().replace(' ', '-').replace('_', '-')
            for sb in software_background + hardware_background
        ]

        # Check if all parts of the marker are covered by user's background
        all_parts_match = True
        for part in parts:
            if not any(part in bg or bg in part for bg in user_backgrounds):
                all_parts_match = False
                break
        if all_parts_match:
            return True

    # Default: don't show marker content
    return False


def adapt_content_by_user_profile(content: str, user_profile: Dict[str, Any]) -> str:
    """
    Adapt content based on user profile using multiple strategies
    """
    # First, sanitize the content for security
    content = sanitize_content(content)

    # Then, handle experience level adaptations
    content = _adapt_by_experience_level(content, user_profile.get('experience_level', 'beginner'))

    # Then, apply marker-based personalization
    content = replace_markers_in_content(content, user_profile)

    # Validate content structure after adaptation
    if not validate_content_structure(content):
        raise ValueError("Content structure is invalid after adaptation")

    return content


def _adapt_by_experience_level(content: str, experience_level: str) -> str:
    """
    Adapt content based on experience level
    """
    if experience_level.lower() == 'beginner':
        # For beginners, add more explanations
        content = content.replace('<p>', '<p class="beginner-focused">', content.count('<p>'))
    elif experience_level.lower() == 'advanced':
        # For advanced users, potentially add more complex examples
        pass  # In a full implementation, we would have more logic here

    return content


def extract_marker_types(content: str) -> List[str]:
    """
    Extract all unique marker types from content
    """
    marker_pattern = r'\{\{personalize:([^}]+)\}\}'
    matches = re.findall(marker_pattern, content)
    return list(set(matches))  # Return unique marker types


def sanitize_content(content: str) -> str:
    """
    Sanitize content to prevent XSS and other security issues
    """
    # This is a more comprehensive implementation
    # For production use, consider using a dedicated library like bleach

    # Remove script tags and their content
    content = re.sub(r'<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>', '', content, flags=re.IGNORECASE | re.DOTALL)

    # Remove style tags with js expressions
    content = re.sub(r'<style[^>]*>.*?<\/style>', '', content, flags=re.IGNORECASE | re.DOTALL)

    # Remove event handlers (like onclick, onload, etc.)
    content = re.sub(r'\s*on\w+\s*=\s*["\'][^"\']*["\']', '', content, flags=re.IGNORECASE)

    # Remove javascript: and vbscript: URIs
    content = re.sub(r'(?i)(javascript:|vbscript:|data:)', '', content)

    # Remove href/src attributes with javascript
    content = re.sub(r'(?i)(href|src)\s*=\s*["\']\s*javascript:[^"\']*["\']', '', content)

    # Remove iframe tags
    content = re.sub(r'<iframe\b[^<]*(?:(?!<\/iframe>)<[^<]*)*<\/iframe>', '', content, flags=re.IGNORECASE | re.DOTALL)

    # Remove object and embed tags
    content = re.sub(r'<(object|embed)\b[^<]*(?:(?!<\/\1>)<[^<]*)*<\/\1>', '', content, flags=re.IGNORECASE | re.DOTALL)

    return content


def validate_content_structure(content: str) -> bool:
    """
    Validate that content markers are properly structured
    """
    import re

    # Find all markers in order
    all_markers = []
    for match in re.finditer(r'\{\{personalize:[^}]+\}\}|{{personalize:end}}', content):
        all_markers.append(match.group())

    # Count and verify proper ordering using a stack-like approach
    depth = 0
    for marker in all_markers:
        if marker.startswith('{{personalize:') and not marker == '{{personalize:end}}':
            # It's a start marker
            depth += 1
        elif marker == '{{personalize:end}}':
            # It's an end marker
            depth -= 1
            if depth < 0:
                return False  # More end markers than start markers at this point

    # All markers should be properly closed
    return depth == 0

File: frontend/src/utils/contentParser.ts
/**
 * Content Parser utility for handling personalization markers in content
 */

/**
 * Interface for personalization marker
 */
export interface PersonalizationMarker {
  type: string;
  content: string;
  startIndex: number;
  endIndex: number;
}

/**
 * Parse content to find personalization markers
 * @param content - The content string to parse
 * @returns Array of PersonalizationMarker objects found in the content
 */
export const findPersonalizationMarkers = (content: string): PersonalizationMarker[] => {
  const markers: PersonalizationMarker[] = [];
  
  // Regular expression to match personalization markers
  // Pattern: {{personalize:markerType}}content{{personalize:end}}
  const markerRegex = /{{personalize:([^}]+)}}([\s\S]*?){{personalize:end}}/g;
  
  let match;
  while ((match = markerRegex.exec(content)) !== null) {
    const fullMatch = match[0];
    const markerType = match[1];
    const markerContent = match[2];
    const startIndex = match.index;
    const endIndex = match.index + fullMatch.length;
    
    markers.push({
      type: markerType,
      content: markerContent,
      startIndex,
      endIndex
    });
  }
  
  return markers;
};

/**
 * Replace personalization markers in content based on user profile
 * @param content - Original content with markers
 * @param userProfile - User's profile information
 * @returns Content with appropriate personalization markers replaced
 */
export const replacePersonalizationMarkers = (
  content: string, 
  userProfile: {
    software_background: string[];
    hardware_background: string[];
    experience_level: string;
  }
): string => {
  // Find all personalization markers
  const markers = findPersonalizationMarkers(content);
  
  // Sort markers in reverse order by position so replacements don't affect indices
  const sortedMarkers = markers.sort((a, b) => b.startIndex - a.startIndex);
  
  let result = content;
  
  for (const marker of sortedMarkers) {
    // Check if this marker should be shown based on user profile
    if (shouldShowMarker(marker.type, userProfile)) {
      // Replace the entire marker with its content
      result = result.substring(0, marker.startIndex) + 
               marker.content + 
               result.substring(marker.endIndex);
    } else {
      // Remove the entire marker (don't show content)
      result = result.substring(0, marker.startIndex) + 
               result.substring(marker.endIndex);
    }
  }
  
  return result;
};

/**
 * Determine if a marker should be shown based on user profile
 * @param markerType - Type of the personalization marker
 * @param userProfile - User's profile information
 * @returns Boolean indicating if marker should be shown
 */
const shouldShowMarker = (
  markerType: string, 
  userProfile: {
    software_background: string[];
    hardware_background: string[];
    experience_level: string;
  }
): boolean => {
  // Special case for experience level markers
  if (markerType === userProfile.experience_level) {
    return true;
  }
  
  // Check if markerType matches any of the user's software backgrounds
  if (userProfile.software_background.some(sb => 
    markerType === sb.toLowerCase().replace(/\s+/g, '-') || 
    markerType === sb.toLowerCase().replace(/\s+/g, '_')
  )) {
    return true;
  }
  
  // Check if markerType matches any of the user's hardware backgrounds
  if (userProfile.hardware_background.some(hb => 
    markerType === hb.toLowerCase().replace(/\s+/g, '-') || 
    markerType === hb.toLowerCase().replace(/\s+/g, '_')
  )) {
    return true;
  }
  
  // If the marker type is a combination (e.g., "python-raspberry-pi"), check if all parts match
  if (markerType.includes('-')) {
    const parts = markerType.split('-');
    const userBgCombined = [
      ...userProfile.software_background.map(sb => sb.toLowerCase().replace(/\s+/g, '-')),
      ...userProfile.hardware_background.map(hb => hb.toLowerCase().replace(/\s+/g, '-'))
    ];
    
    return parts.every(part => 
      userBgCombined.some(ubg => ubg.includes(part) || part.includes(ubg))
    );
  }
  
  // If no specific match found, don't show the content
  return false;
};

/**
 * Extract all unique marker types from content
 * @param content - The content string to analyze
 * @returns Array of unique marker types found in the content
 */
export const extractMarkerTypes = (content: string): string[] => {
  const markers = findPersonalizationMarkers(content);
  const uniqueTypes = new Set(markers.map(marker => marker.type));
  return Array.from(uniqueTypes);
};

/**
 * Remove all personalization markers from content, keeping only the generic content
 * @param content - The content string with markers
 * @returns Content with all personalization markers removed
 */
export const removeMarkers = (content: string): string => {
  const markers = findPersonalizationMarkers(content);
  
  // Sort markers in reverse order by position so replacements don't affect indices
  const sortedMarkers = markers.sort((a, b) => b.startIndex - a.startIndex);
  
  let result = content;
  
  for (const marker of sortedMarkers) {
    // Remove the entire marker
    result = result.substring(0, marker.startIndex) + 
             result.substring(marker.endIndex);
  }
  
  return result;
};